src/NUnitFramework/framework/Internal/AttributeProviderWrapper.cs

```diff
@@ -52,14 +51,27 @@ public object[] GetCustomAttributes(bool inherit)
 
         public bool IsDefined(Type attributeType, bool inherit)
         {
-            return GetCustomAttributes(attributeType, inherit).Any();
+            return GetCustomAttributes(attributeType, inherit).Length > 0;
         }
 
-        private static T[] GetFiltered(IEnumerable<object> attributes)
+        private static T[] GetFiltered(object[] attributes)
         {
-            return attributes
-                   .OfType<T>()
-                   .ToArray();
+            List<T> filtered = null;
+            foreach (var attribute in attributes)
+            {
+                if (attribute is T t)
+                {
+                    filtered ??= new List<T>();
+                    filtered.Add(t);
+                }
+            }
+
+            return filtered?.ToArray() ??
+#if NETSTANDARD2_0_OR_GREATER
+                Array.Empty<T>();
+#else
+                new T[0];
+#endif
```

Comment:
When glancing over the code I really got confused  about this change, because from a readibility point of view it is definitely worse than before 😞 Took me quite a few brain cycles to see that it just replaces the linq usage and that it is correct (which it is, as far as I can tell!).

So at least I would have expected a seperate commit doing this "get rid of linq" change,  accompanied with the motivating performance numbers in the commit message telling WHY this  decrease in readability is needed, e.g. because it improves the performance. Would make it easier to swallow that bitter pill here ;-)
Right now it is embedded in the "Cache method metadata" commit, which this change above has nothing to do with.

Please take this comment with a grain of salt, as this is just a "drive by review", as I am not a contributer to this project and therefore I don't think my opinion really matters here...

Reply:
Fair enough, optimizations tend to get a bit verbose and harder to read. "Cache method metadata" might not tell the store, but I put it under the "Reduce allocations" sub-heading under PR 😉 So it was one of the things that contributed to performance improvements, I should have included the rough allocation numbers in hindsight though.

This method was called quite a lot and it usually returned empty array at the end.  It was hot w.r.t to allocations so it was part of the cleanup for the biggest offenders. I might create a separate PR for other long-hanging fruit but this already got a bit out of hand... 

Thanks for the drive-by, all feedback is always appreciated!